### ISSUE TITLE
Property & fuzz tests of the advanced spend API

### ISSUE THREAD

**Issue by gz-c** [skycoin/skycoin#1251]
Easier to test this at the wallet.Service layer

Use https://golang.org/pkg/testing/quick/ or some other property tester.

Also look into fuzzing it, if that could produce different results.

**Comment by gz-c** [comment]
Testing `wallet.Service.CreateAndSignTransactionAdvanced` can be complicated.

Start with testing `wallet.ChooseSpends` first, it is easier to do property testing for.

**Comment by Vavilen** [comment]
PR #1567

**Comment by Vavilen** [comment]
@gz-c , is it correct?
https://github.com/skycoin/skycoin/blob/a5b688a94ceefbc19a797a7b3d0a11bb3f658c31/src/wallet/wallet_test.go#L1844


Here we break the []UxBalance collecting loop if got enough coins. So, filtered spending []UxBalance length with zero hours can be less, than all zero []UxBalance.
https://github.com/skycoin/skycoin/blob/a5b688a94ceefbc19a797a7b3d0a11bb3f658c31/src/wallet/wallet.go#L1791-L1800

**Comment by gz-c** [comment]
It is adding up outputs that don't have hours (the `zero` array), then it checks if hours are sufficient again after the break.  Then it adds outputs that have hours (`nonzero` array) as necessary.

**Comment by Vavilen** [comment]
@gz-c , yes and it means that it is not always correct
https://github.com/skycoin/skycoin/blob/a5b688a94ceefbc19a797a7b3d0a11bb3f658c31/src/wallet/wallet_test.go#L1844
, because collected spending `haveZero` count, which is summary zero hours UxBalances count [here](https://github.com/skycoin/skycoin/blob/8888c4e83e54686c2f7442cf4646393b4c6303a8/src/wallet/wallet_test.go#L1750),  can be bigger, than `len(zeroBalances)`, if we got enough in [loop](https://github.com/skycoin/skycoin/blob/a5b688a94ceefbc19a797a7b3d0a11bb3f658c31/src/wallet/wallet.go#L1791-L1800) end finished with `break`.

**Comment by gz-c** [comment]
@Vavilen can you give me example input that shows the problem

**Comment by Vavilen** [comment]
@gz-c ,in #1567  I added temporary `hours` param (it was hardcoded as 0) for TestWalletChooseSpendsMaximizeUxOuts and TestWalletChooseSpendsMinimizeUxOutsRandom. 
You can run it and see the result. In most of cases you can see the result like:
```console
=== RUN   TestWalletChooseSpendsMinimizeUxOutsRandom
--- FAIL: TestWalletChooseSpendsMinimizeUxOutsRandom (0.00s)
	Error Trace:	wallet_test.go:2038
			wallet_test.go:1417
	Error:      	Not equal: 
	            	expected: 26
	            	actual: 1
FAIL
```

**Comment by gz-c** [comment]
Save the input that causes this bug, and share it here.  We can add it to the unit test to find the bug problem.  I need to see the inputs that are triggering the bug to understand it.

**Comment by Vavilen** [comment]
@gz-c , I added TestWalletChooseSpendsMaximizeUxOutsFailedCase to the wallet/wallet_test.go with the input data which ls probably triggering the bug.
Please, look at https://github.com/skycoin/skycoin/pull/1567/files#diff-9421da5c2d94d12c10ade6afd06dd7bfR1410

**Comment by iketheadore** [comment]
@Vavilen It checks that the `haveZero` equal to `len(zeroBalances)` only when `len(nonzeroBalances) > 0`
https://github.com/skycoin/skycoin/blob/a5b688a94ceefbc19a797a7b3d0a11bb3f658c31/src/wallet/wallet_test.go#L1842-L1845

The `ChooseSpend` method won't collect none zero hour UxBalance until all zero hour UxBalance are collected, so if the collection loop breaks before all zero hour UxBalance are collected, the `len(nonezeroBalances)` must be `0` and the check won't be trigged.

**Comment by Vavilen** [comment]
@iketheadore The `ChooseSpend` method can stop collecting `zero` hour UxBalance if there is enough **coins** collected. But it will continue collect from `nonzero` if it is not enough **hours** collected. So, it is possible when `len(nonzeroBalances) > 0` but `haveZero` and  `len(zeroBalances)` are not equal.
Here is an example with failing test data: https://github.com/skycoin/skycoin/pull/1567/files#diff-9421da5c2d94d12c10ade6afd06dd7bfR1410
Please, let me know, if I'm wrong.

**Comment by iketheadore** [comment]
@Vavilen yes, you're right, it's possible that the `coins` are enough but `hours` are not enough, in this case it will try to choose uxouts from `nonezero`, and the `zero` might not be all collected.
